### PR TITLE
Short-circuit ByteBuf::release

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:

ReferenceCountUtil::safeRelease can both hit interface virtual calls and requires checking for an interface type (ReferenceCounted)

Modifications:

Perform a class check to save both.

Result:

Faster buffers release

### CODE REVIEW

**Comment by franz1981** [comment]
This just remove the `itable_stub` presence on release, on the right, eg

image

Despite being a tiny amount, overall, in the global picture, for the write/flush path, is still an unecessary cost, see

image

where the `itable_stub` cost ~1.5%, a tiny fraction but as costy as performing useful work.

**Comment by hyperxpro** [comment]
In these scenarios, I really love pattern-type matches.

File: transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java [hunk 1]
```diff
@@ -276,8 +277,19 @@ public boolean remove() {
...
+            // and makes better use of virtual calls vs interface ones
+            if (msg instanceof AbstractReferenceCountedByteBuf) {
+                try {
+                    // release now as it is flushed.
+                    ((AbstractReferenceCountedByteBuf) msg).release();
+                } catch (Throwable t) {
+                    logger.warn("Failed to release a ByteBuf: {}", msg, t);
+                }
+            } else {
+                // only release message, notify and decrement if it was not canceled before.
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java:290]: this comment should be moved as it applies for both cases

Comment (franz1981) [reply transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java:290]: You meant the log one?

Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/ChannelOutboundBuffer.java:290]: No... `// only release message, notify and decrement if it was not canceled before.` Just move it above the `if (!e.cancelled) {`

**Review by chrisvest — APPROVED** [review]

**Pull request merged**